File: LC/Python/Medium/236.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':        
        def dfs(curr):
            if not curr or curr == p or curr == q:
                return curr
            
            l = dfs(curr.left)
            r = dfs(curr.right)
            
            if l and r:
                return curr
            
            return l or r
            
        
        return dfs(root)

File: LC/C++/Medium/236.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
     TreeNode* dfs(TreeNode* curr, TreeNode* p, TreeNode* q) {
        if (curr == nullptr || curr == p || curr == q) {
            return curr;
        }

        TreeNode* l = dfs(curr->left, p, q);
        TreeNode* r = dfs(curr->right, p, q);

        if (l && r) {
            return curr;
        }

        return l != nullptr ? l : r;
     }
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        return dfs(root, p, q);
    }
};